sql/catalog: fix policy dependency handling during backup/restore

spilchen opened the pull request:
Previously, when performing table level restores, policies with missing dependencies (UDFs, types, or relations) were not handled correctly. Unlike triggers, which are dropped when their dependencies are missing, policies would remain in an inconsistent state, potentially causing restore failures or leaving dangling references.

This change extends the existing trigger dependency handling logic to also cover RLS policies. When a policy depends on objects that aren't being restored (e.g., a UDF referenced in a USING clause), the policy is now properly dropped. The implementation also ensures that back-references are cleaned up appropriately, preventing dangling references that could interfere with subsequent operations like dropping sequences or types.

Additionally, the schema changer state rewriting logic is enhanced to handle missing dependencies for policy expressions (PolicyUsingExpr, PolicyWithCheckExpr, PolicyDeps) during restore operations.

Fixes #151042.

Release note (bug fix): Fixed an issue where Row Level Security policies with missing dependencies during table-level restores could cause inconsistent state or restore failures.

Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151045)
<!-- Reviewable:end -->

rafiss reviewed (approved):
nice job finding this bug!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151045#-:-OWvuCo_4XfV4BMtr9nx:bend7p3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/testdata/end_to_end/create_policy_with_deps/create_policy_with_deps.definition` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151045#-OWvtzAS5Dw3ArfLrjXO:-OWvtzAS5Dw3ArfLrjXP:b-4yqd64) ([raw file](https://github.com/cockroachdb/cockroach/blob/18ccceaed32ae76eed5b738d1baaae834e4e9c94/pkg/sql/schemachanger/testdata/end_to_end/create_policy_with_deps/create_policy_with_deps.definition#L11)):*
> ```
> 
> test
> CREATE POLICY admin_policy ON users 
> ```

super nit: if the only reason for adding this test was to make sure backup/restore was handled correctly, then it might be better to add a more specialized test case in `pkg/backup/testdata/backup-restore`, and skip adding this one here.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151045)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @rafiss)

___
*[`pkg/sql/schemachanger/testdata/end_to_end/create_policy_with_deps/create_policy_with_deps.definition` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151045#-OWvtzAS5Dw3ArfLrjXO:-OX-IpWS2ioJNo2Azo0u:brvg0xw) ([raw file](https://github.com/cockroachdb/cockroach/blob/18ccceaed32ae76eed5b738d1baaae834e4e9c94/pkg/sql/schemachanger/testdata/end_to_end/create_policy_with_deps/create_policy_with_deps.definition#L11)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: if the only reason for adding this test was to make sure backup/restore was handled correctly, then it might be better to add a more specialized test case in `pkg/backup/testdata/backup-restore`, and skip adding this one here.
</blockquote></details>

That was the only reason. I added the test under `pkg/backup/testdata/backup-restore` now.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/catalog: fix policy dependency handling during backup/restore

Previously, when performing table level restores, policies with missing
dependencies (UDFs, types, or relations) were not handled correctly.
Unlike triggers, which are dropped when their dependencies are missing,
policies would remain in an inconsistent state, potentially causing
restore failures or leaving dangling references.

This change extends the existing trigger dependency handling logic to
also cover RLS policies. When a policy depends on objects that aren't
being restored (e.g., a UDF referenced in a USING clause), the policy is
now properly dropped. The implementation also ensures that
back-references are cleaned up appropriately, preventing dangling
references that could interfere with subsequent operations like dropping
sequences or types.

Additionally, the schema changer state rewriting logic is enhanced to
handle missing dependencies for policy expressions (PolicyUsingExpr,
PolicyWithCheckExpr, PolicyDeps) during restore operations.

Fixes #151042.

Release note (bug fix): Fixed an issue where Row Level Security policies
with missing dependencies during table-level restores could cause
inconsistent state or restore failures.

Epic: None

spilchen commented:
TFTR!

bors r+
